04-onion-routing.md

```diff
@@ -1052,25 +1055,62 @@ The erring node then generates a new key, using the key type `ammag`.
 This key is then used to generate a pseudo-random stream, which is in turn
 applied to the packet using `XOR`.
 
-The obfuscation step is repeated by every hop along the return path.
-Upon receiving a return packet, each hop generates its `ammag`, generates the
-pseudo-random byte stream, and applies the result to the return packet before
-return-forwarding it.
+When supported, the erring node should also populate the `attribution_data` field in `update_fail_htlc` consisting of the following data:
 
-The origin node is able to detect that it's the intended final recipient of the
-return message, because of course, it was the originator of the corresponding
-forward packet.
-When an origin node receives an error message matching a transfer it initiated
-(i.e. it cannot return-forward the error any further) it generates the `ammag`
-and `um` keys for each hop in the route.
-It then iteratively decrypts the error message, using each hop's `ammag`
-key, and computes the HMAC, using each hop's `um` key.
-The origin node can detect the sender of the error message by matching the
-`hmac` field with the computed HMAC.
+1. data:
+    * [`20*u32`:`htlc_hold_times`]
+    * [`210*sha256[..4]`:`truncated_hmacs`]
 
-The association between the forward and return packets is handled outside of
-this onion routing protocol, e.g. via association with an HTLC in a payment
-channel.
+The field `htlc_hold_times` contains the htlc hold time in milliseconds for each hop. The sender can use this information to score nodes on latency. Nodes along the path that lack accurate timing information may simply report a value of zero. In such cases, the sender should distribute any potential latency penalty across multiple nodes. This encourages path nodes to provide timing data to avoid being held responsible for the high latency of other nodes.
```

Comment:
We could clarify how the hold time is computed exactly. Do we start counting when receiving `update_add_htlc`, `commitment_signed` or `revoke_and_ack`? It should probably be `revoke_and_ack` because that's when we can start relaying.

Reply:
I think it is up to implementations to decide how to compute hold time exactly, and make it so that it is in the best interest of their users. In the end, the rational thing to do is look at the sender logic for penalizing latency, and optimize for minimal penalty?